### ISSUE TITLE
[ENHANCEMENT] Better Support for YAML Frontmatter Keywords

### ISSUE THREAD

**Issue by canpolat** [Zettlr/Zettlr#1092]
## Description
There are (at least) two separate ways to add some metadata to a document with Zettlr:

1. YAML frontmatter (YFM)
2. Adding tags using `#`

In a lot of the cases I have to add both YFM (title) and `#`-tags. I think having the possibility to merge these two could be nice. The main idea is supporting a YFM variable (for example, `tags`) for the same purpose as the `#`-tags in a document. Zettlr already parses YFM (to extract `title`), this will build on that and add parsing of `tags`. If tags are merged into YFM, I would expect Zettlr to behave as follows…

## Example Workflow
The user would have a YFM that looks like this:

```
---
title: Very interesting document
id: MyZettlrId
tags:
    - tag1
    - tag2
---
```

I would expect Zettlr to:

1. Auto-complete the YFM tags
2. Make the YFM tags clickable (related to #607)
3. Parse the tags that are present in YFM and add them to tag cloud
4. Still support internal linking via `[[#mytag]]` format  (with the `#`)

The tags can be added to a document either by using `#` or via the YFM without the `#` (this is to avoid problems with them looking like comments).

## Caveats
The documentation ([YAML Frontmatter](https://docs.zettlr.com/en/core/yaml-frontmatter/)) does not list `tags` as a supported variable, so I suppose adding it should not be a problem, but if it conflicts with some other use of tags, that could be a problem (and may require creation of a unique variable name).

This may also be confusing for beginners unless it is well documented. But, as long as both ways of handling tags have the same functionality, I don't see this as a huge risk.

As a corner case scenario, a user may add tags to both YFM and the content. But I guess that should not behave any different than the user adding the same tag twice in a document.

YFM supports spaces in tags (via the use of quotation marks). This may be a problem for Zettlr. I don't know the best solution for this scenario but I suppose one of the following can be applied:

1. The offending tag can be marked with squiggly lines and the use may be warned about the situation
2. The offending tag can be auto-corrected to a version without spaces

**Comment by boutil** [comment]
@canpolat, isn't what you want provided by the `keywords` entry in the YAML frontmatter?
See the corresponding [paragraph in the docs](https://docs.zettlr.com/en/core/yaml-frontmatter/#variable-keywords).

**Comment by canpolat** [comment]
@boutil Not really.

First of all, `keywords` are used by pandoc to insert metadata to the exported PDF document. I don't want my internal tags to be visible from outside.

Also:
- Keywords are not auto-completed
- Keywords are not clickable (as far as I can see)
- Contrary to documentation, they do not appear in the tag cloud (maybe this is a bug?)

**Comment by nathanlesage** [comment]
> YFM supports spaces in tags (via the use of quotation marks). This may be a problem for Zettlr.

Absolutely not :) 

Furthermore, as @canpolat and @boutil have rightfully stated is that the right variable is "keywords", not "tags" (this is Pandoc specific, but this way your keywords/tags should be compatible across the whole eco-system). Zettlr will treat the array of `keywords` as tags.

Concerning that they're not clickable yet, I have actually no quick fix for this, but given enough time one might figure out a way to make Zettlr recognize them as well. Auto-completion is also difficult. (Technical background: We'd need a pretty sophisticated formatter, which is capable (a) of reading YAML, and (b) able to see that you make use of the keywords-variable…)

> Contrary to documentation, they do not appear in the tag cloud (maybe this is a bug?)

Umm … they should appear there. So possibly a bug o.O

But I think there are some more things off with the tag cloud, so if you have some time, try to collect some more issues with the tag cloud so that we have a better way of quickly fixing them …

**Comment by canpolat** [comment]
> >    Contrary to documentation, they do not appear in the tag cloud (maybe this is a bug?)
>
> Umm … they should appear there. So possibly a bug o.O
>
> But I think there are some more things off with the tag cloud, so if you have some time, try to collect some more issues with the tag cloud so that we have a better way of quickly fixing them …

I dug a bit deeper and I think I found the problem: I was using the `tab` character for indentation and `tab` is not allowed in YAML. Zettlr, indeed, shows squiggly lines under the `tab` to indicate that there is a problem (see screenshot 1). But I think I tried using keywords only once and upon not seeing the keyword in the tag cloud I gave up. I will create a PR for documentation to draw attention to this.

The only problem I can see with the tag cloud is the function-less close buttons appearing next to the tag names (screenshot 2).

**Screenshot 1**
![image](https://user-images.githubusercontent.com/459891/88034665-4e963100-cb41-11ea-8edf-1060d15992d0.png)

**Screenshot 2**
![image](https://user-images.githubusercontent.com/459891/88035267-2fe46a00-cb42-11ea-951b-92dc71ffd59b.png)

**Comment by nathanlesage** [comment]
>  I was using the tab character for indentation and tab is not allowed in YAML.

Ah, yes, I noticed this today as well, and I was like "why are you adding f***ing Tabs for this", need to get to this

> The only problem I can see with the tag cloud is the function-less close buttons appearing next to the tag names (screenshot 2).

Function-less? Oh, wait you mean the "1x"-thingy? Well, looks like a UX issue b/c you obviously can't "close" the tag, it's the amount of times it appears within the files …

**Comment by canpolat** [comment]
> Function-less? Oh, wait you mean the "1x"-thingy? Well, looks like a UX issue b/c you obviously can't "close" the tag, it's the amount of times it appears within the files …

Oh! I thought "1" as "number of times" and "X" as "close". Now that you say it, it is obvious. I thought "the close button" was a left over from days where it has some function (remove a tag from some place) and now no longer tied to any logic :D

**Comment by nathanlesage** [comment]
> I thought "the close button" was a left over from days where it has some function (remove a tag from some place) and now no longer tied to any logic :D

… and I now see why you would think that :D 

Welcome to the bright world of UX design 🌈

**Comment by nathanlesage** [comment]
Short update, as the stalebot has hit this issue:

1. Consolidation of YAML frontmatter variables will be performed according to the JAMS-standard, currently under development; Zettlr will incorporate that standard asap, hence: no `tags`-property, if there is no standard for it.
2. With regard to autocomplete of the YAML frontmatter: I'll get to this automatically as soon as we have some autocomplete for the YAML properties (as both are functionally the same)

So I'll be closing this issue, because said functionality is a MUST on the Zettlr roadmap (+ to keep the issue list clean, as always, yey)

**Comment by canpolat** [comment]
1. Not having the `tags` is not a problem at all. I can live with `keywords` 👍 
2. I'm looking forward to the YAML auto-complete feature. That's going to be legendary 🥇

**Comment by nathanlesage** [comment]
Yeah, let's hope for the best, but depending on how the progress is, I'll want to rewrite A LOT of the underlying logic either way for CodeMirror.next, so maybe I'll postpone it till then ;)

**Comment by Bitgug** [comment]
Is Zettlr mobile for Android supposed to recognize YAML 'keywords' as tags, too? Because it doesn't with me. It only recognizes YAML 'tags' as tags. Whereas Zettlr desktop for Windows does.

**Comment by nathanlesage** [comment]
I am pretty sure you are on the wrong repository. There is no Zettlr mobile.

**Comment by Bitgug** [comment]
How very inattentive of me. My apologies. Thank you (and for Zettlr desktop)!
